smithyctl/command/component/package.go

```diff
@@ -0,0 +1,208 @@
+
+	"github.com/smithy-security/smithyctl/internal/registry"
+)
+
+var flags packageFlags
+
+type (
+	packageFlags struct {
+		specPath               string
+		packageVersion         string
+		sdkVersion             string
+		registryURL            string
+		registryBaseRepository string
+		registryAuthEnabled    bool
+		registryAuthUsername   string
+		registryAuthPassword   string
+	}
+)
+
+// NewPackageCommand returns a new package command.
+func NewPackageCommand() *cobra.Command {
+	cmd := &cobra.Command{
+		Use:   "package",
+		Short: "Packages a component's configuration",
+		RunE: func(cmd *cobra.Command, args []string) error {
+			if err := packageComponent(cmd.Context(), flags); err != nil {
+				return errors.Errorf("unexpected failure: %w", err)
+			}
+			return nil
+		},
+	}
+
+	cmd.
+		Flags().
+		StringVar(
+			&flags.specPath,
+			"spec-path",
+			".",
+			"the location of the component spec file",
+		)
+	cmd.
+		Flags().
+		StringVar(
+			&flags.packageVersion,
+			"version",
+			"",
+			"the version to be used to package the component",
+		)
+	_ = cmd.MarkFlagRequired("version")
+	cmd.
+		Flags().
+		StringVar(
+			&flags.sdkVersion,
+			"sdk-version",
+			"",
+			"the version the sdk used to build the component",
+		)
+	_ = cmd.MarkFlagRequired("sdk-version")
+	cmd.
+		Flags().
+		StringVar(
+			&flags.registryURL,
+			"registry-url",
+			"ghcr.io",
+			"the reference to the OCI compliant registry",
+		)
+	cmd.
+		Flags().
+		StringVar(
+			&flags.registryBaseRepository,
+			"registry-base-repository",
+			"smithy-security/manifests/components",
+			"the repository where to push manifests to",
+		)
+	cmd.
+		Flags().
+		BoolVar(
+			&flags.registryAuthEnabled,
+			"registry-auth-enabled",
+			false,
+			"if enabled, it requires authentication for the registry",
+		)
+	cmd.
+		Flags().
+		StringVar(
+			&flags.registryAuthUsername,
+			"registry-auth-username",
+			"",
+			"the username used for authentication for the registry",
+		)
+	cmd.
+		Flags().
+		StringVar(
+			&flags.registryAuthPassword,
+			"registry-auth-password",
+			"",
+			"the password used for authentication for the registry",
+		)
+
+	return cmd
+}
+
+func packageComponent(ctx context.Context, flags packageFlags) error {
+	component, err := parseComponentSpec(flags.specPath)
+	if err != nil {
+		return err
+	}
+
+	reg, err := registry.New(
+		flags.registryURL,
+		flags.registryBaseRepository,
+		flags.registryAuthEnabled,
+		flags.registryAuthUsername,
+		flags.registryAuthPassword,
+	)
+	if err != nil {
+		return errors.Errorf("failed to initialize registry: %w", err)
+	}
+
+	if err := reg.Package(ctx, registry.PackageRequest{
+		Component:        component,
+		SDKVersion:       flags.sdkVersion,
+		ComponentVersion: flags.packageVersion,
+	}); err != nil {
+		return errors.Errorf("failed to package component: %w", err)
+	}
+
+	return nil
+}
+
+func parseComponentSpec(path string) (*v1.Component, error) {
+	const (
```

Comment:
this looks like it could have side effects by forgotten package files in the root of the repo, I would rather we keep it simple and check if the path exists, is a file and if it can be parsed. in any other case we just return a clear error and the user should figure out what they did wrong

Reply:
My idea is that in general it's nice to have this at root level and just rely on default behaviour

This is often the case with lots of tools - like wrangler etc

I think that we can pass a full path, potentially - but if you feel strongly I can change it - I don't mind

I do think that the proposal is better in terms of UX though!

Reply:
I ended up using the file directly for simplicity